source/misc/bash_startup.in

```diff
@@ -1,257 +1,95 @@
-    iterm2_begin_osc
-    printf "1337;RemoteHost=%s@%s" "$USER" "$iterm2_hostname"
-    iterm2_end_osc
-
-    iterm2_begin_osc
-    printf "1337;CurrentDir=%s" "$PWD"
-    iterm2_end_osc
-
-    iterm2_print_user_vars
-  }
-
-  # Usage: iterm2_set_user_var key value
-  function iterm2_set_user_var() {
-    iterm2_begin_osc
-    printf "1337;SetUserVar=%s=%s" "$1" $(printf "%s" "$2" | base64)
-    iterm2_end_osc
-  }
-
-  # Users can write their own version of this method. It should call
-  # iterm2_set_user_var but not produce any other output.
-  function iterm2_print_user_vars() {
-    true
-  }
-
-  function iterm2_prompt_prefix() {
-    iterm2_begin_osc
-    printf "133;D;\$?"
-    iterm2_end_osc
-
-    iterm2_print_state_data
-
-    iterm2_begin_osc
-    printf "133;A"
-    iterm2_end_osc
-  }
-
-  function iterm2_prompt_suffix() {
-    iterm2_begin_osc
-    printf "133;B"
-    iterm2_end_osc
-  }
-
-  function iterm2_print_version_number() {
-    iterm2_begin_osc
-    printf "1337;ShellIntegrationVersion=1"
-    iterm2_end_osc
-  }
-
-
-  # If hostname -f is slow on your system, set iterm2_hostname before sourcing this script.
-  if [ -z "$iterm2_hostname" ]; then
-    iterm2_hostname=$(hostname -f)
-  fi
-  iterm2_preexec_install
-
-  # This is necessary so the first command line will have a hostname and current directory.
+iterm2_prompt_prefix () {
+  printf $osc'133;D;'$?$cso
   iterm2_print_state_data
-  iterm2_print_version_number
-fi
+  printf $osc'133;A'$cso
+}
+
+iterm2_prompt_suffix() {
+  printf $osc'133;B'$cso
+}
+
+preexec() {
+  printf $osc'133;C'$cso
+}
+
+# This function is installed as bash PROMPT_COMMAND; invoked before each
+# interactive prompt display. It sets preexec_interactive_mode for DEBUG trap.
+iterm2_preexec_invoke_cmd() {
+  status=$?
+  last_command="$(echo $(fc -ln | tail -1))"; (exit $status) # Reset $? back.
+
+  PS1="\[$(iterm2_prompt_prefix)\]$orig_ps1\[$(iterm2_prompt_suffix)\]"
+  prev_ps1=$PS1; (exit $status) # Save PS1 in case user changes PS1; reset $?.
+
+  preexec_interactive_mode=yes # This must be the last line in this function.
+}
+
+# This function is the DEBUG trap. It inspects the current environment, and
+# if it is ready to be invoked interactively, it 'preexec's the command line.
+iterm2_preexec_invoke_exec() {
+  [ ! -t 1 ] && return # We're in a piped subshell (STDOUT is not a TTY).
+  [ -n "$COMP_LINE" ] && return    # We're in the middle of a completion.
+  [ -z "$preexec_interactive_mode" ] && return # We're displaying prompt.
+  [ -"$BASH_SUBSHELL"- = -"0"- ] && preexec_interactive_mode=
+  [ -"iterm2_preexec_invoke_cmd"- = -"$BASH_COMMAND"- ] &&
+                              preexec_interactive_mode= && return
+
+  fc=$(echo $(fc -ln | tail -1)); prev_fc=$last_fc; last_fc=$fc
+  [ x"$prev_fc" != x"$fc" ] && this_command=$fc || this_command=
+
+  preexec "$this_command"
+}
+
+iterm2_preexec_install() {
+  [ -z "$BASH" ] && set -o functrace  >/dev/null 2>&1 &&
```

Comment:
Precisely. Again, I'm trying to make some of this reusable for other shells. I think it would be ideal if there weren't a need for a separate file for other Almquist/Korn compatible shells.